ride/swap.ride

```diff
@@ -0,0 +1,277 @@
+
+func getStringOrFailFromAddress(address: Address, key: String) = address.getString(key).valueOrErrorMessage(makeString(["mandatory ", address.toString(), ".", key, " is not defined"], ""))
+
+let keyFeeCollectorAddress = "%s__feeCollectorAddress"
+let feeCollectorAddress = factoryContract.getStringOrFailFromAddress(keyFeeCollectorAddress).addressFromStringValue()
+
+
+
+
+
+func asInt(val: Any) = {
+  match val {
+    case valInt: Int => valInt
+    case _ => throw("fail to cast into Int")
+  }
+}
+
+func dataMappingPoolAssets(internalAmountAssetStr: Int, internalPriceAssetStr: Int) = {
+  makeString(
+    ["%d%d",
+    internalAmountAssetStr.toString(),
+    internalPriceAssetStr.toString()],
+    SEP)
+}
+
+func managerPublicKeyOrUnit() = match keyManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func pendingManagerPublicKeyOrUnit() = match keyPendingManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func isManager(i: Invocation) = {
+  match managerPublicKeyOrUnit() {
+    case pk: ByteVector => i.callerPublicKey == pk
+    case _: Unit => i.caller == this
+  }
+}
+
+func mustManager(i: Invocation) = {
+  i.isManager() || "permission denied".throw()
+}
+
+func getAccBalance(assetId: String) = {
+  if(assetId == "WAVES") then wavesBalance(this).available else assetBalance(this, fromBase58String(assetId))
+}
+
+@Callable(i)
+func swapCalculateREADONLY(
+  amountIn: Int,
+  assetIn: String,
+  assetOut: String
+) = {
+  let prFee = getInteger(protocolFee()).value()
+  let feeProtocolAmountCalc = fraction(amountIn.toBigInt(), prFee.toBigInt(), feeScale)
+  let plFee = getInteger(poolFee()).value()
+  let feePoolAmountCalc = fraction(amountIn.toBigInt(), plFee.toBigInt(), feeScale)
+
+  let feeProtocolAmount = if (feeProtocolAmountCalc.toInt() == 0) then {
+    toBigInt(1)
+  } else {
+    feeProtocolAmountCalc
+  }
+  let feePoolAmount = if (feePoolAmountCalc.toInt() == 0) then {
+    toBigInt(1)
+  } else {
+    feePoolAmountCalc
+  }
+  let cleanAmountIn = amountIn.toBigInt() - feeProtocolAmount - feePoolAmount
+
+  let lpNonReverse = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetIn, assetOut],
+    []
+  )
+
+  let lpReverse = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetOut, assetIn],
+    []
+  )
+
+  let (isReverse, pool) = if (lpNonReverse != unit) then {
+    let isReverse = false
+    let pool = invoke(
+      factoryContract,
+      "getPoolAddressFromLpAssetREADONLY",
+      [lpNonReverse],
+      []
+    ).exactAs[String]
+    (isReverse, pool)
+  } else if (lpReverse != unit) then {
+    let isReverse = true
+    let pool = invoke(
+      factoryContract,
+      "getPoolAddressFromLpAssetREADONLY",
+      [lpReverse],
+      []
+    ).exactAs[String]
+    (isReverse, pool)
+  } else {
+    poolNotExist
+  }
```

Comment:
```suggestion
  let (isReverse, lpAssetIdStr) = if (lpNonReverse != unit) then {
    (false, lpNonReverse.value())
  } else if (lpReverse != unit) then {
    (true, lpReverse.value())
  } else {
    poolNotExist
  }
  let pool = factoryContract.invoke(
    "getPoolAddressFromLpAssetREADONLY",
    [lpAssetIdStr],
    []
  ).exactAs[String]
```

Reply:
partly agree. but you stiil need call "getPoolAddressFromLpAssetREADONLY" to get pool address, because lpReverse.value() is lp asset but not pool address

Reply:
this invocation is on the 122nd line, isn't it?